### PR TITLE
Vector Index Maintenance

### PR DESCRIPTION
**Pull request by lkuffo**
## Features

- Adding maintenance (`INSERT`, `DELETE`) and concurrency control to PDXearch index.   
   - Commits stage row ids and tombstone deletes.
   - First search fetches the committed rows back and builds each new row group's index once.
- Mirroring DuckDB physical rowgroups via `table.GetRowGroupCollection()->GetRowGroups()` (needs DuckDB > `1.5.1`). 
- Parquet ingestion and checkpointing (merging/dropping rowgroups) are supported.
- For now, we will start pointing to `duckir-integration` in PDX.

## Pending (before merging)
- Cross check code
- Clean unit tests

<details>

<summary><h3>Maintenance design</h3></summary>

* `Append` / `Insert` only stage row ids, independently of their source.
    * Note that we don't have rowgroup boundaries here. This is the whole reason we must defer indexing until scan.
* Reconciliation only at Scan:
    * We lock [shared] the index and check (`SyncAndLockForSearch`):
        * Are there staged row id ranges?
        * Mismatch between DuckDB and our layout? (to detect rg merging)
        * (`!IsInSyncWithTable`)
    * If out of sync, Then run `SyncWithTable()` under an exclusive lock
    * Why not at the next insert?
        * Because a bulk load call `Append` many times 
        * and Current commit’s rows are not placed yet
* `SyncWithTable`:
    * Detect and rebuild rowgroups whose mirrors no longer match 
        * E.g., due to checkpoint merging two rowgroups
        * `FindStaleRowGroups()` → Detects rg merges
    * Index staged rows
        * We gather all staged row chunks per physical rowgroup
        * A rowgroup with no mirror is k-means indexed
        * Otherwise, we Append to existing index
* How do we fetch rows from IDs?
    * `RowGroupCollection::Fetch` committed rows visibility (skipping NULLs)
* Downsides:
    * New rows are scanned twice (one row group at a time)
    * First scan after bulk insert pays the indexing price.
    * Sync holds an exclusive lock 

</details>

<details>

<summary><h3>Unit tests</h3></summary> 

  * Inserts and Deletes 
  * Rolled-back insert, insert-then-delete (same transaction), multi-statement transactions
  * Merge of two rowgroups
  * Unaligned rowgroups (non full); Parquet ingestion
  * Transaction isolation
  * Concurrent maintenance (testing locks are doing its job)
  * Failed commit
  * Two indexes in the same table
  * ALTER TABLE (only ADD COLUMN is possible)
  * VACUUM
  * Degenerate ones:
      * Index table with 1 row then all inserts
      * Merging NULL-only rowgroup
      * One commit in a rowgroup boundary
      * 3 partial rowgroups merge into two
      * Delete all rows, checkpoint, reinsert
      * Rows deleted before their first Sync (before index scan)

</details>

### CODE REVIEW

File: src/include/index/pdxearch_index.hpp [hunk 1]
```diff
@@ -30,6 +37,29 @@ class PDXearchIndex : public BoundIndex {
...
+
+	// Row ids committed to the table whose embeddings are not in the index yet
+	std::vector<PDXearchRowRange> unindexed_row_ranges;
+	std::atomic<bool> has_unindexed_rows {false};
+
+	void AppendRow(idx_t row_group_idx, row_t row_id, const float *transformed_embedding);
+	void DeleteRow(row_t row_id);
+
+	// Returns the committed rows of [start, end) whose embedding is not NULL,
+	// We transforme them (random rotation) and returns how many were written to `embeddings`.
```
Comment (Noorts) [inline src/include/index/pdxearch_index.hpp]: nit:

```suggestion
	// We transform them (random rotation) and return how many were written to `embeddings`.
```

File: src/index/pdxearch_index.cpp [hunk 2]
```diff
@@ -58,75 +72,334 @@ PDXearchIndex::PDXearchIndex(const string &name, IndexConstraintType index_const
...
  * Index maintenance
  ******************************************************************/
 
 ErrorData PDXearchIndex::Append(IndexLock &lock, DataChunk &entries, Vector &row_ids) {
-	// Execute all column expressions before inserting the data chunk.
-	DataChunk expr_chunk;
-	expr_chunk.Initialize(Allocator::DefaultAllocator(), logical_types);
-	ExecuteExpressions(entries, expr_chunk);
-	// Now insert the data chunk.
-	return Insert(lock, expr_chunk, row_ids);
```
Comment (Noorts) [inline src/index/pdxearch_index.cpp]: Are we sure we want to column expression code here? I don't know what it does or whether we need it but duckdb-vss does [still use it](https://github.com/duckdb/duckdb-vss/blob/2e566dac6e9f0064aea8b6e834357189183d5b0b/src/hnsw/hnsw_index.cpp#L571) (though it might be stale).

File: src/index/pdxearch_index.cpp [hunk 3]
```diff
@@ -58,75 +72,334 @@ PDXearchIndex::PDXearchIndex(const string &name, IndexConstraintType index_const
...
 
+// Called by DuckDB at commit with the final row ids, before the rows enter the table's row groups. Which row group
+// they land in is not knowable here, so only the ids are recorded; SyncWithTable indexes them once the table knows.
 ErrorData PDXearchIndex::Insert(IndexLock &lock, DataChunk &data, Vector &row_ids) {
 	auto _lock = rwlock.GetExclusiveLock();
 
-	throw NotImplementedException("PDXearchIndex::Insert() not implemented");
+	const idx_t count = data.size();
+	if (count == 0) {
+		return ErrorData();
```
Comment (Noorts) [inline src/index/pdxearch_index.cpp]: nit: would it make sense to use the `ErrorData("error message here")` constructor to clarify the error?

File: src/index/pdxearch_index.cpp [hunk 4]
```diff
@@ -58,75 +72,334 @@ PDXearchIndex::PDXearchIndex(const string &name, IndexConstraintType index_const
...
+	}
+	row_ids.Flatten(count);
+	const auto row_id_data = FlatVector::GetData<row_t>(row_ids);
+
```
Comment (Noorts) [inline src/index/pdxearch_index.cpp]: nit: It seems we're building consecutive ranges. To improve readability it might help to add a simple comment, such as the following.

```suggestion
    // Determine consecutive ranges of row ids.
    // For example: `1, 2, 3, 5, 6 --> [1, 3] + [5, 6]`.
```

File: src/index/pdxearch_index.cpp [hunk 5]
```diff
@@ -12,11 +18,19 @@ PDXearchIndex::PDXearchIndex(const string &name, IndexConstraintType index_const
...
 		throw NotImplementedException("PDXearch indexes do not support unique or primary key constraints");
 	}
+	if (persistence_info.IsValid()) {
+		throw InternalException(
+		    "Something went wrong: the PDXearch index was created without its table. This is likely because a "
+		    "malformed persisted index was loaded. Index persistence is not supported yet, but DuckDB will still "
+		    "try to persist it. Open your database file manually (duckdb test.db) and drop the index(es). Run "
+		    "'SELECT sql FROM duckdb_indexes();' to see the indexes, and then 'DROP INDEX index_name;' to drop the "
+		    "unused index(es).");
+	}
```
Comment (Noorts) [inline src/index/pdxearch_index.cpp]: I'm not sure if this catches the same error case this internal exception used to but it's fine by me. We can enhance persistence safety and corner cases when we implement persistence.

**Review by Noorts — APPROVED** [review]
LGTM!

Read the code; skimmed the tests. Overall looking good, given the constraints we're dealing with. Very nice to take Parquet, checkpointing, and dynamic row group size into account as well.

Incredible improvement for the extension, thanks! Looking forward to what's next.
